cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -59,49 +66,144 @@ func (c *healthCheck) Name() string {
-		name:   "StaticPodManifest",
-		client: client,
-		f:      staticPodManifestHealth,
-	})
-
 	return preflight.RunChecks(healthChecks, os.Stderr, ignoreChecksErrors)
 }
 
-// apiServerHealthy checks whether the API server's /healthz endpoint is healthy
-func apiServerHealthy(client clientset.Interface) error {
-	healthStatus := 0
+// CreateJob is a check that verifies that a Job can be created in the cluster
+func createJob(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration) (lastError error) {
+	jobName := "upgrade-health-check"
+	ns := metav1.NamespaceSystem
+	timeoutSeconds := 15
 
-	// If client.Discovery().RESTClient() is nil, the fake client is used, and that means we are dry-running. Just proceed
-	if client.Discovery().RESTClient() == nil {
-		return nil
+	// Prepare Job
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      jobName,
+			Namespace: ns,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit:          utilpointer.Int32Ptr(0),
+			ActiveDeadlineSeconds: utilpointer.Int64Ptr(int64(timeoutSeconds)),
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyNever,
+					SecurityContext: &v1.PodSecurityContext{
+						RunAsUser:    utilpointer.Int64Ptr(999),
+						RunAsGroup:   utilpointer.Int64Ptr(999),
+						RunAsNonRoot: utilpointer.BoolPtr(true),
+					},
+					Tolerations: []v1.Toleration{
+						{
+							Key:    "node-role.kubernetes.io/master",
+							Effect: v1.TaintEffectNoSchedule,
+						},
+					},
+					Containers: []v1.Container{
+						{
+							Name:  jobName,
+							Image: images.GetPauseImage(cfg),
+							Args:  []string{"-v"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Check if the Job already exists and delete it
+	if _, err := client.BatchV1().Jobs(ns).Get(jobName, metav1.GetOptions{}); err == nil {
+		if err = deleteHealthCheckJob(client, ns, jobName); err != nil {
+			return err
+		}
+	}
+
+	// Cleanup the Job on exit
+	defer func() {
+		lastError = deleteHealthCheckJob(client, ns, jobName)
+	}()
+
+	timeout := time.Second * time.Duration(timeoutSeconds)
+
+	// Create the Job, but retry in case it is being currently deleted
+	klog.V(2).Infof("Creating Job %q in the namespace %q", jobName, ns)
+	err := wait.PollImmediate(time.Second*1, timeout, func() (bool, error) {
+		if _, err := client.BatchV1().Jobs(ns).Create(job); err != nil {
+			klog.V(2).Infof("Could not create Job %q in the namespace %q, retrying: %v", jobName, ns, err)
+			lastError = err
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return errors.Wrapf(lastError, "could not create Job %q in the namespace %q", jobName, ns)
+	}
+
+	// Waiting and manually deleteing the Job is a workaround to not enabling the TTL controller.
+	// TODO: refactor this if the TTL controller is enabled in kubeadm once it goes Beta.
+
+	// Wait for the Job to complete
+	err = wait.PollImmediate(time.Second*1, timeout, func() (bool, error) {
```

Comment:
As long as this is done by `kubeadm upgrade apply` I am fine with it, but if this is to be called by other actions as well, it would be necessary to convert it to a watch.